libdnf5/base/goal.cpp

```diff
@@ -2206,6 +2412,130 @@ GoalProblem Goal::Impl::add_reason_change_to_goal(
+        // For reverted action INSTALL and reason CLEAN the previous reason could have been either DEPENDENCY or WEAK DEPENDENCY
+        // to pick the right one we have to look into history.
+        if ((package_replay.action == Action::REASON_CHANGE) ||
+            (package_replay.action == Action::INSTALL && pkg.get_reason() == Reason::CLEAN)) {
+            // We look up the reason based on only name and arch, this means we could find a different
+            // version of installonly package however we store only one reason for ALL versions of
+            // installonly packages so it doesn't matter.
+            package_replay.reason =
+                history->transaction_item_reason_at(pkg.get_name(), pkg.get_arch(), reverting_transaction.get_id() - 1);
+        } else if (
+            package_replay.action == Action::REMOVE &&
+            (pkg.get_reason() == Reason::DEPENDENCY || pkg.get_reason() == Reason::WEAK_DEPENDENCY)) {
+            package_replay.reason = Reason::CLEAN;
+        } else {
+            package_replay.reason = pkg.get_reason();
+        }
+
+        replay.packages.push_back(package_replay);
+    }
+
+    for (const auto & group : reverting_transaction.get_comps_groups()) {
+        transaction::GroupReplay group_replay;
+        group_replay.group_id = group.to_string();
+        // Do not revert UPGRADE for groups. Groups don't have an upgrade path so they cannot be
+        // upgraded or downgraded. The UPGRADE action is basically a synchronization with
+        // current group definition. Reverting synchronization is again synchronization but
+        // with the older group definition, this happens automatically by reverting the rpm
+        // actions. We only have to keep the UPGRADE action to have a record of the operation.
+        if (group.get_action() != transaction::TransactionItemAction::UPGRADE) {
+            auto reverted_action = REVERT_ACTION.find(group.get_action());
+            if (reverted_action == REVERT_ACTION.end()) {
+                libdnf_throw_assertion(
+                    "Cannot revert action: \"{}\"", transaction_item_action_to_string(group.get_action()));
+            }
+            group_replay.action = reverted_action->second;
+        } else {
+            group_replay.action = transaction::TransactionItemAction::UPGRADE;
+        }
+
+        if (group_replay.action == Action::INSTALL && group.get_reason() == Reason::CLEAN) {
+            group_replay.reason = Reason::DEPENDENCY;
+        } else if (group_replay.action == Action::REMOVE && group.get_reason() == Reason::DEPENDENCY) {
+            group_replay.reason = Reason::CLEAN;
+        } else {
+            group_replay.reason = group.get_reason();
+        }
+
+        // Don't fill in @System repo_id because it's not possible to perform forward actions from it
+        if (group.get_repoid() != "@System") {
```

Comment:
I believe that this is the same type of problem like with RPMs. What about to not use repo_id for undo?